cms/admin.py

```diff
@@ -1,12 +1,164 @@
 '''Base classes for the CMS admin interface.'''
 from django.contrib import admin
 from django.conf import settings
+from django.core.urlresolvers import NoReverseMatch, reverse
+from django.db.models.deletion import get_candidate_relations_to_delete
+from django.db.utils import DEFAULT_DB_ALIAS
 from reversion.admin import VersionAdmin
 from watson.admin import SearchAdmin
 
+from cms.apps.pages.models import Page
 from cms.models.base import PageBaseSearchAdapter, SearchMetaBaseSearchAdapter
 
 
+def check_inline_for_admin_url(obj, inline, parent, inline_check=True):
+    '''
+    A function used to get the admin URL for an obj.
+
+    Takes the object itself, an inline object and the class of the
+    parent model.
+
+    If no change URL is found, None is returned.
+    '''
+    # We obviously need the object to be the same type of object
+    # as the inline's parent to at least get anywhere useful. We
+    # catch thisstraight away and immediately return if so to save
+    # time.
+
+    if inline_check and not isinstance(obj, inline.model):
+        return None
+
+    # We've got an inline for this model. Lets check the fk_name
+    # attribute on the InlineModelAdmin. If it's set we'll use
+    # that, else we'll find the ForeignKey to the parent.
+
+    obj_fk_name = getattr(inline, 'fk_name', False)
+    if obj_fk_name:
+        # the fk_name value is set so we assume that it will
+        # resolve a parent since an inline can't be saved without
+        # a parent.
+
+        obj_parent = getattr(obj, obj_fk_name, False)
+        try:
+            return reverse(
+                f'admin:{obj_parent._meta.app_label}_{obj_parent._meta.model_name}_change',
+                args=[obj_parent.pk]
+            )
+        except NoReverseMatch:
+            pass
+
+    # If we make it here, the fk_name attribute was not set or the
+    # parent did not resolve to an object We'll now look for
+    # specifically ForeignKeys to the parent model. There can't be
+    # more than one as if there were, Django would throw errors on
+    # inline registration due to the lack of the fk_name
+    # attribute to distinguish the two fields.
+
+    for field in obj._meta.get_fields():
+        if field.get_internal_type() in ['ForeignKey', 'OneToOneField']:
+            # Follow the ForeignKey to find the related model.
+
+            related_model = obj._meta.get_field(field.attname).rel.to
+
+            if parent == related_model:
+                # We've found a Foreign key to the parent, now
+                # to extract the page and get its admin change URL.
+
+                try:
+                    field_value = getattr(obj, field.attname)
+                except AttributeError:
+                    field_value = None
+
+                if field_value:
+                    try:
```

Comment:
Is this a "can't happen" try/catch? If it's guarding against something that should never happen, it should fail loudly.

Reply:
(Same applies to the one above it.)